fv/wireguard_test.go

```diff
@@ -253,6 +255,81 @@ var _ = infrastructure.DatastoreDescribe("WireGuard-Supported", []apiconfig.Data
 	})
 
 	Context("traffic with Wireguard enabled", func() {
+		// Checks the TCP dump for a count value. Retries until count is correct, or fails after 1.5s.
+		waitForPackets := func(t *tcpdump.TCPDump, timeout time.Time, name string, num int) error {
+			for ;; time.Now().Before(timeout) {
+				if num == 0 && t.MatchCount(name) > 0 {
+					// We expect no traffic, but got some.  Error immediately.
+					break
+				}
+				if t.MatchCount(name) >= num {
+					// We expected some packets and have got at least the required number (we allow more for handshake
+					// etc.)
+					return nil
+				}
+				time.Sleep(100 * time.Millisecond)
+			}
+			return fmt.Errorf("incorrect packet count for %s; expected=%d actual=%d", name, num, t.MatchCount(name))
+		}
+
+		// Runs wg and extracts the received and sent packet counts.
+		getWgStatistics := func(felix *infrastructure.Felix) (sent, rcvd string) {
+			xferRegExp := regexp.MustCompile(`transfer:\s+([0-9a-zA-Z. ]+)\s+received,\s+([0-9a-zA-Z. ]+)\s+sent`)
+			out, err := felix.ExecOutput("wg")
+			Expect(err).NotTo(HaveOccurred())
+			matches := xferRegExp.FindStringSubmatch(out)
+			if len(matches) != 3 {
+				return
+			}
+			return matches[2], matches[1]
+		}
+
+		// Checks connectivity between workloads 0 and 1, checking that eth0 traffic is via wireguard port and that
+		// there are non-empty wireguard stats.
+		checkConn := func() error {
+			// Reset TCP packet counts.
+			By("Resetting the TCP dump counts")
+			for i := range felixes {
+				tcpdumps[i].ResetCount("numInTunnelPackets")
+				tcpdumps[i].ResetCount("numOutTunnelPackets")
+				tcpdumps[i].ResetCount("numWorkload0to1Packets")
+				tcpdumps[i].ResetCount("numWorkload1to0Packets")
+			}
+
+			// Send packets to and from workloads on each felix.
+			By("Sending packets W1->W2 and W2->W1")
+			if err, _ := wls[0].SendPacketsTo(wls[1].IP, 5, 56); err != nil {
+				return err
+			}
+			if err, _ := wls[1].SendPacketsTo(wls[0].IP, 5, 56); err != nil {
+				return err
+			}
+
+			// Now check the packet counts are as expected. We should have no WL->WL traffic visible on eth0, but
+			// we should be able to see tunnel traffic. Since we want to verify
+			By("Checking the packet stats from tcpdump")
+			timeout := time.Now().Add(2*time.Second)
+			for i := range felixes {
+				if err := waitForPackets(tcpdumps[i], timeout, "numInTunnelPackets", 10); err != nil {
+					return err
+				} else if err := waitForPackets(tcpdumps[i], timeout, "numOutTunnelPackets", 10); err != nil {
+					return err
+				} else if err := waitForPackets(tcpdumps[i], timeout, "numWorkload0to1Packets", 0); err != nil {
+					return err
+				} else if err := waitForPackets(tcpdumps[i], timeout, "numWorkload1to0Packets", 0); err != nil {
+					return err
+				}
+			}
+
+			By("Checking the packet stats from wg")
+			for i := range felixes {
+				rcvd, sent := getWgStatistics(felixes[i])
+				Expect(rcvd).NotTo(BeEmpty())
+				Expect(sent).NotTo(BeEmpty())
```

Comment:
@robbrockbank - why are we removing sent/rcvd comparison check?

Reply:
At the moment I'm trying to reduce test flakiness.  It's possible I've been a little overzealous, but at the moment I haven't convinced myself yet that the sent and received packets should always align pm either side - in particular during establishment of the routing on both sides.

I think I'd rather err on the side of caution and just check that "yes we have some traffic" and that we are hitting the correct ports on the host interface.